beacon-chain/rpc/beacon_server.go

```diff
@@ -136,6 +147,113 @@ func (bs *BeaconServer) BlockTree(ctx context.Context, _ *ptypes.Empty) (*pb.Blo
 	}, nil
 }
 
+// eth1Data is a mechanism used by block proposers vote on a recent Ethereum 1.0 block hash and an
+// associated deposit root found in the Ethereum 1.0 deposit contract. When consensus is formed,
+// state.latest_eth1_data is updated, and validator deposits up to this root can be processed.
+// The deposit root can be calculated by calling the get_deposit_root() function of
+// the deposit contract using the post-state of the block hash.
+func (bs *BeaconServer) eth1Data(ctx context.Context) (*pbp2p.Eth1Data, error) {
+	beaconState, err := bs.beaconDB.HeadState(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("could not fetch beacon state: %v", err)
+	}
+	// Fetch the current canonical chain height from the eth1.0 chain.
+	currentHeight := bs.powChainService.LatestBlockHeight()
+	eth1FollowDistance := int64(params.BeaconConfig().Eth1FollowDistance)
+	stateLatestEth1Hash := bytesutil.ToBytes32(beaconState.LatestEth1Data.DepositRoot)
+	// If latest ETH1 block hash is empty, send a default response
+	if stateLatestEth1Hash == [32]byte{} {
+		return bs.defaultEth1DataResponse(ctx, currentHeight, eth1FollowDistance)
+	}
+	// Fetch the height of the block pointed to by the beacon state's latest_eth1_data.block_hash
+	// in the canonical eth1.0 chain.
+	_, stateLatestEth1Height, err := bs.powChainService.BlockExists(ctx, stateLatestEth1Hash)
+	if err != nil {
+		return nil, fmt.Errorf("could not verify block with hash exists in Eth1 chain: %#x: %v", stateLatestEth1Hash, err)
+	}
+	dataVotes := []*pbp2p.Eth1Data{}
+	voteCountMap := make(map[string]voteHierarchy)
+	bestVoteHeight := big.NewInt(0)
+	depositCount, depositRootAtHeight := bs.beaconDB.DepositsNumberAndRootAtHeight(ctx, currentHeight)
+	var mostVotes uint64
+	var bestVoteHash string
+	for _, vote := range beaconState.Eth1DataVotes {
+		if ctx.Err() != nil {
+			return nil, ctx.Err()
+		}
+		// Let dataVotes be the set of Eth1DataVote objects vote in state.eth1_data_votes where:
```

Comment:
This comment block is confusing and it feels like this section of logic should be abstracted into its own method.

What is going on in this loop? Abstracting this would make the intent clear and make this piece of code easier to test.

Reply:
abstracted away into isValidVote()